core/drivers/atmel_saic.c

```diff
@@ -0,0 +1,356 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <assert.h>
+#include <drivers/atmel_saic.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <kernel/interrupt.h>
+#include <kernel/pm.h>
+#include <libfdt.h>
+#include <sam_sfr.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+#define AT91_AIC_MAX_PRIO	8
+
+#define SAMA5D2_AIC_MAX_IRQS	77
+
+#define SAMA5D2_AIC_MAX_IRQS32	((SAMA5D2_AIC_MAX_IRQS + 31) / 32)
+
+struct saic_data {
+	struct itr_chip chip;
+	vaddr_t base;
+	size_t nr_irqs;
+	uint32_t external[SAMA5D2_AIC_MAX_IRQS32];
+};
+
+static struct saic_data saic = {0};
+
+static void saic_register_pm(void);
+
+static void saic_write_reg(uint32_t reg, uint32_t val)
+{
+	io_write32(saic.base + reg, val);
+}
+
+static uint32_t saic_read_reg(uint32_t reg)
+{
+	return io_read32(saic.base + reg);
+}
+
+void atmel_saic_it_handle(void)
+{
+	uint32_t irqnr = saic_read_reg(AT91_AIC_IVR);
+
+	itr_handle(irqnr);
+	saic_write_reg(AT91_AIC_EOICR, 0);
+}
+
+static void saic_select_it(size_t it)
+{
+	assert(!(it & ~AT91_AIC_SSR_ITSEL_MASK));
+
+	saic_write_reg(AT91_AIC_SSR, it);
+}
+
+static void saic_configure_it(size_t it, int src_type, int priority)
```

Comment:
I would have expected `uint32_t` instead of `int` here.
Any particular reason why `int` is used here?

Reply:
No, you are right, using `uint32_t` would be better.

Reply:
I wil lalso change the `it` parameter of `saic_get_src_type()` to use a `size_t`